pin/pin.go

```diff
@@ -159,22 +160,46 @@ func (p *pinner) isInternalPin(key key.Key) bool {
 func (p *pinner) IsPinned(k key.Key) (string, bool, error) {
 	p.lock.RLock()
 	defer p.lock.RUnlock()
-	return p.isPinned(k)
+	return p.isPinnedWithType(k, "all")
 }
 
-// isPinned is the implementation of IsPinned that does not lock.
+func (p *pinner) IsPinnedWithType(k key.Key, typeStr string) (string, bool, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	return p.isPinnedWithType(k, typeStr)
+}
+
+// isPinnedWithType is the implementation of IsPinnedWithType that does not lock.
 // intended for use by other pinned methods that already take locks
-func (p *pinner) isPinned(k key.Key) (string, bool, error) {
-	if p.recursePin.HasKey(k) {
+func (p *pinner) isPinnedWithType(k key.Key, typeStr string) (string, bool, error) {
+	switch typeStr {
+	case "all", "direct", "indirect", "recursive", "internal":
+	default:
+		err := fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, internal, all}", typeStr)
+		return "", false, err
+	}
+	if (typeStr == "recursive" || typeStr == "all") && p.recursePin.HasKey(k) {
 		return "recursive", true, nil
 	}
-	if p.directPin.HasKey(k) {
+	if typeStr == "recursive" {
+		return "", false, nil
+	}
+
+	if (typeStr == "direct" || typeStr == "all") && p.directPin.HasKey(k) {
 		return "direct", true, nil
 	}
-	if p.isInternalPin(k) {
+	if typeStr == "direct" {
+		return "", false, nil
+	}
+
+	if (typeStr == "internal" || typeStr == "all") && p.isInternalPin(k) {
 		return "internal", true, nil
 	}
+	if typeStr == "internal" {
+		return "", false, nil
+	}
```

Comment:
Here if something is pinned both directly and recursively, we just answer that it is pinned recursively. The changes in this PR don't change this answer. The same answer was given previously.

If you are worried about that it's probably better to create a separate issue.